### PR TITLE
Extract SystemTableProvider from Hive

### PR DESCRIPTION
**Pull request by Dith3r**
## Description

Allow to define system table providers for Iceberg similar to Hive.

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RawSystemTableProvider.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    public Optional<SchemaTableName> getSourceTableName(SchemaTableName table)
+    {
+        try {
+            String name = tableNameFrom(table.getTableName());
+            return Optional.of(new SchemaTableName(table.getSchemaName(), name));
+        }
+        catch (IllegalArgumentException e) {
+            return Optional.empty();
+        }
+    }
```
Comment (ksobolew) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RawSystemTableProvider.java:57]: I'm not exactly sure how that interface works - once we derive the base name, is `getSystemTable` called with the *original* or *derived* table name?

Comment (Dith3r) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/RawSystemTableProvider.java:57]: `getSystemTable` is called with original `SchemaTableName`. `getSourceTableName` is used for access checks (at least in hive).

**Comment by Dith3r** [comment]
CI failure: https://github.com/trinodb/trino/issues/22861

File: core/trino-spi/src/main/java/io/trino/spi/connector/SystemTableProvider.java [hunk 2]
```diff
@@ -11,17 +11,13 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package io.trino.plugin.hive;
-
-import io.trino.spi.connector.ConnectorSession;
-import io.trino.spi.connector.SchemaTableName;
-import io.trino.spi.connector.SystemTable;
+package io.trino.spi.connector;
```
Comment (raunaqmorarka) [inline core/trino-spi/src/main/java/io/trino/spi/connector/SystemTableProvider.java:14]: What's the benefit of putting this in SPI ?
You can either put it in trino-plugin-toolkit or just make a copy in iceberg

Comment (Dith3r) [reply core/trino-spi/src/main/java/io/trino/spi/connector/SystemTableProvider.java:14]: > or just make a copy in iceberg

Copy will not allow making `SystemTableProviders` which could be used with more than one connector.

> You can either put it in trino-plugin-toolkit

I will not argue if SPI is a better place, but I placed `SystemTableProvider` in the same place where `SystemTable` is defined.

Comment (Dith3r) [reply core/trino-spi/src/main/java/io/trino/spi/connector/SystemTableProvider.java]: `SystemTableProvider` moved to `trino-plugin-toolkit`

**Review by raunaqmorarka — APPROVED** [review]

**Review by wendigo — APPROVED** [review]

**Comment by electrum** [comment]
What is the reasoning for this change? Do you have plans to use this for other connectors? If we don't have any near-term plans, then let's revert this, and it can be added when it will be used.

**Comment by electrum** [comment]
Also, there is a cost to sharing code, and seemingly no benefit here. What's the usefulness in extracting an interface that will be used independently by different connectors? There is no shared implementation that would need to be fixed in multiple places. But if we need to make a change to one of the connectors but not the others, the artificial coupling here makes that change harder.

If you need a similar internal interface in another connector, simply copy the (two method) interface to the other connector.

**Pull request merged**